guardian/pkg/asyncutil/async.go

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) 2025 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package asyncutil
+
+import (
+	"context"
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+)
+
+type CommandExecutor[C any, R any] interface {
+	Send(C) <-chan Result[R]
+	PauseExecution()
+	ResumeExecution()
+	ShutdownSignaler() Signaler
+}
+
+type CommandDispatcher []interface {
+	PauseExecution()
+	ResumeExecution()
+	ShutdownSignaler() Signaler
+}
+
+func (dispatcher CommandDispatcher) PauseExecution() {
+	for _, executor := range dispatcher {
+		executor.PauseExecution()
+	}
+}
+
+func (dispatcher CommandDispatcher) ResumeExecution() {
+	for _, executor := range dispatcher {
+		executor.ResumeExecution()
+	}
+}
+
+func (dispatcher CommandDispatcher) WaitForShutdown() {
+	for _, executor := range dispatcher {
+		<-executor.ShutdownSignaler().Receive()
+	}
+}
+
+type commandExecutor[Req any, Resp any] struct {
+	command             func(context.Context, Req) (Resp, error)
+	pauseExecution      chan chan struct{}
+	resumeExecution     Signaler
+	cmdChan             chan Command[Req, Resp]
+	backlogChan         chan Command[Req, Resp]
+	inflightCmds        sync.WaitGroup
+	executeSig          Signaler
+	shutdownCompleteSig Signaler
+	errBuff             AsyncErrorBuffer
+}
+
+// NewCommandExecutor creates a new RequestHandler implementation.
+func NewCommandExecutor[C any, R any](ctx context.Context, errBuff AsyncErrorBuffer, f func(context.Context, C) (R, error)) CommandExecutor[C, R] {
+	hdlr := &commandExecutor[C, R]{
+		command:         f,
+		errBuff:         errBuff,
+		cmdChan:         make(chan Command[C, R], 100),
+		backlogChan:     make(chan Command[C, R], 100),
+		pauseExecution:  make(chan chan struct{}),
+		resumeExecution: NewSignaler(),
+
+		executeSig:          NewSignaler(),
+		shutdownCompleteSig: NewSignaler(),
+	}
+
+	go hdlr.loop(ctx)
+	return hdlr
+}
+
+func (executor *commandExecutor[C, R]) Receive() <-chan Command[C, R] {
+	return executor.cmdChan
+}
+
+func (executor *commandExecutor[C, R]) ShutdownSignaler() Signaler {
+	return executor.shutdownCompleteSig
+}
+
+func (executor *commandExecutor[C, R]) Send(params C) <-chan Result[R] {
+	cmd, resultChan := NewCommand[C, R](params)
+	executor.cmdChan <- cmd
```

Comment:
Can we send 101 requests ?

Reply:
Not at the same time, but the other side of the channel shouldn't block so there wouldn't be any issue.